### PR TITLE
Improve accusation flow & logging information for slashing event

### PR DESCRIPTION
**Pull request by lorenzo-dev1**
Contributions:
- reduces number of raised accusations by inferring additional information from committed blocks 
- to facilitate timely mining of innocence proofs, sets tip of accountability tx to 1 wei (so that they get gossiped to the network)
- fixes a bug in PVN rule in fault detector
- removes PVO3 rule, it is now caught at proposal decoding time
- fixes a bug in the accusation verifier precompiled (proofs with votes for nil should be rejected)
- improves logging for slashing event to facilitate debugging by node operator
- limit accusations raised over a single height to `maxAccusationPerHeight` per validator per height (i.e. cannot accuse the same validator with more than `maxAccusationPerHeight` accusations over the same height).
- general code quality refactor and test additions

### CODE REVIEW

File: autonity/types.go [hunk 1]
```diff
@@ -27,6 +26,55 @@ const (
...
+			"1. had precommitted in vr for a different value.\n" +
+			"2. had precommitted for a non-nil value between vr and currentRound.\n" +
+			"3. there is a quorum of prevotes for a value different than the proposed one at vr.\n" +
+			"4. If accusation, the local validator was not able to provide a quorum of prevotes at vr for the proposed value.\n" +
+			"Reference: line 15, 36 and 40 in " + paper //nolint
+	case PVN:
+		explanation = "Validator has broadcasted a prevote for a proposal with vr == -1, but he either:\n" +
+			"1. was locked on a different value.\n" +
+			"2. If accusation, the local validator was not able to provide the proposal corresponding to his prevote.\n" +
+			"NOTE: this could happen also for a prevote for a proposal with vr != 1.\n" +
```
Comment (vpiyush) [inline autonity/types.go:48]: correction => vr != -1

File: autonity/types.go [hunk 2]
```diff
@@ -27,6 +26,55 @@ const (
 	Innocence
 )
 
+// human understandable explanation of accountability rules
+func (r Rule) Explanation() string {
+	paper := "page 6 of \"The latest gossip on BFT consensus\" by Ethan Buchman, Jae Kwon, and Zarko Milosevic"
```
Comment (vpiyush) [inline autonity/types.go:31]: should we give a link here?

File: autonity/types.go (hunk 1 above)
Comment (lorenzo-dev1) [reply autonity/types.go]: Thanks! fixed

File: autonity/types.go (hunk 2 above)
Comment (lorenzo-dev1) [reply autonity/types.go]: Good idea, I added it

File: autonity/autonity.go [hunk 3]
```diff
@@ -201,25 +205,28 @@ func (c *AutonityContract) MinimumBaseFee(block *types.Header, db vm.StateDB) (*
...
 	if !ok {
-		p := c.electProposer(header, height, round)
-		roundMap = make(map[int64]common.Address)
-		roundMap[round] = p
-		c.proposers[height] = roundMap
-		// since the growing of blockchain height, so most of the case we can use the latest buffered height to gc the buffer.
-		c.trimProposerCache(height)
-		return p
+		c.proposers[height] = make(map[int64]common.Address)
+		needsTrimming = true // cannot trim here, we might delete the entry we just created if it is for a very old height
```
Comment (vpiyush) [inline autonity/autonity.go:217]: We should return here, this can cause panic roundMap is nil and being accessed later

Comment (lorenzo-dev1) [reply autonity/autonity.go:217]: good catch, this should fix it https://github.com/autonity/autonity/pull/937/commits/25d95f2838bd9905a8f35cd7433b42cebd64aa6c

**Review by vpiyush — APPROVED** [review]
LGTM!

**Review by Jason-Zhangxin-Chen — APPROVED** [review]
LGTM, left few comments.

File: consensus/tendermint/accountability/fault_detector.go [hunk 4]
```diff
@@ -945,57 +977,57 @@ func (fd *FaultDetector) newPrevotesAccountabilityCheck(height uint64, prevote m
...
+	* - latest non-nil precommit is equivocated
+	 */
 	return nil
 }
 
 func (fd *FaultDetector) oldPrevotesAccountabilityCheck(height uint64, quorum *big.Int, correspondingProposal *message.Propose, prevote message.Msg) (proof *Proof) {
 	currentR := correspondingProposal.R()
 	validRound := correspondingProposal.ValidRound()
 
-	// If one prevotes for an invalid old proposal, then it should be a misbehaviour.
```
[the remark this answers is not in the document]
Comment (lorenzo-dev1) [reply consensus/tendermint/accountability/fault_detector.go:987]: yes we do discard these proposals at decoding time now.

**Review by yazzaoui — APPROVED** [review]
LGTM, great work

**Pull request merged**